### DISCUSSION TITLE
【微前端样式】求助主应用antd5，子应用antd4 样式问题

### DISCUSSION THREAD

**Discussion by consistent-k** [ant-design/ant-design#43368] [Q&A] [answered]
1、主应用因技术栈升级，将antd4 升级至5，使用umi4 + qiankun + antd5。
2、子应用已经稳定上线，使用umi4 + qianku + antd4。
3、开启qiankun的样式隔离策略。
4、子应用中的弹出组件，如Dropdown组件 下拉框样式无法加载，主应用升级antd5之前展示正常，（猜测是因为主应用内部有antd4样式）
5、求助：目前除了给Dropdown等组件添加getPopupContainer 修改弹出框定位不渲染到Body这种做法以外，还有没有更优解？

**Comment by arvinxx** [comment]
主应用prefix 改下呗？比如从改成 ant 改成 ant5，这样就不冲突了。

**Reply by consistent-k** [reply]
现在的问题是子应用 antd4的组件渲染到了body上，body上没有antd4的样式，改主应用的prefix解决不了这个问题吧

**Reply by yoyo837** [reply]
啥antd4样式不样式的,不都默认是ant-*吗?

**Reply by consistent-k** [reply]
嗯？那我现在使用antd5不是废弃掉之前那个单独的css文件了吗?  不引入这个文件不就没有antd4 全部组件的样式吗，antd5的样式是 in js的吧

**Reply by consistent-k** [reply]
比如这个  ant-dropdown 是子应用的下拉菜单，渲染到了主应用body上，主应用的antd5并没有这个样式了
![image](https://github.com/ant-design/ant-design/assets/67736912/1afb4284-70bb-4532-b674-036476528267)

**Reply by MadCcc** [reply]
antd v4 的样式还是需要引入 css 文件哈，这个是全局的

**Reply by consistent-k** [reply]
主应用已经是antd5了，这种情况下怎么引入antd4的样式是个问题吧，假设我复制了一份antd4的css文件引入，是否会对antd5的组件造成冲突呢。

**Reply by afc163** [reply]
这个确实是个坑，目前只能用 getPopupContainer。更好的方式感觉需要 qiankun 来解。 cc @kuitos

**Reply by kuitos** [reply]
运行时处理会碰到各种各样的corner case，大多也需要组件库配合，而且有额外的性能开销。现在qiankun的策略基本都是推荐用户配置组件库前缀来解决，不过目前antd还是有些全局样式前缀没覆盖上，上次跟闲夕聊过，可能要找个机会一起处理一下

**Reply by arvinxx** [reply, marked as answer]
> 现在的问题是子应用 antd4的组件渲染到了body上，body上没有antd4的样式，改主应用的prefix解决不了这个问题吧

想到一种骚操作看看是否可行：

1. 主应用使用 antd v5 ，并设置自定义前缀，比如为 master-antd5；
2. 主应用引入一份 antd v4 css；

由于主应用的antd 前缀改了，得益于运行时 cssinjs，主应用下的antd组件前缀也自动改为 master-antd5，因此这些组件不会受到 antd v4 css 的影响。同时主应用因为引入了 antd v4 的css，因此主应用里也有了一份 ant 前缀的样式表；

这个时候子应用由于还是 antd v4，因此哪怕 antd4 的组件挂载在 body 上，因为主应用挂载了 antd 4 样式，样式也会是正常的

**Reply by consistent-k** [reply]
这样应该可以解决问题，我先这样试一下。另外应该可以引入单个组件的样式吧，其实渲染在body上的元素应该就DropDown、Modal、Select等组件，遇到问题的组件单独引入引入样式，这样可以避免引入全量css导致产出变大的问题吧。
